src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectSystem/ProjectConfigurationStateManager.cs

```diff
@@ -0,0 +1,167 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.ProjectSystem;
+using Microsoft.AspNetCore.Razor.Serialization;
+using Microsoft.AspNetCore.Razor.Utilities;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.Logging;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor.Utilities;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
+
+/// <summary>
+/// Updates project system service with new project info.
+/// </summary>
+/// <remarks>
+/// Used to figure out if the project system data is being added, updated or deleted
+/// and acts accordingly to update project system service.
+/// </remarks>
+internal partial class ProjectConfigurationStateManager : IDisposable
+{
+    private readonly ProjectSnapshotManagerDispatcher _projectSnapshotManagerDispatcher;
+    private readonly IRazorProjectService _projectService;
+    private readonly IProjectSnapshotManager _projectManager;
+    private readonly ILogger _logger;
+
+    private readonly AsyncBatchingWorkQueue<(ProjectKey ProjectKey, RazorProjectInfo? ProjectInfo)> _workQueue;
+    private readonly CancellationTokenSource _disposalTokenSource;
+
+    public ProjectConfigurationStateManager(
+        ProjectSnapshotManagerDispatcher projectSnapshotManagerDispatcher,
+        IRazorProjectService projectService,
+        ILoggerFactory loggerFactory,
+        IProjectSnapshotManager projectManager)
+    {
+        _projectSnapshotManagerDispatcher = projectSnapshotManagerDispatcher;
+        _projectService = projectService;
+        _projectManager = projectManager;
+        _logger = loggerFactory.GetOrCreateLogger<ProjectConfigurationStateManager>();
+
+        _disposalTokenSource = new ();
+        _workQueue = new(
+            EnqueueDelay,
+            ProcessBatchAsync,
+            _disposalTokenSource.Token);
+    }
+
+    private ValueTask ProcessBatchAsync(ImmutableArray<(ProjectKey ProjectKey, RazorProjectInfo? ProjectInfo)> workItems, CancellationToken cancellationToken)
+    {
+        foreach (var workItem in workItems.GetMostRecentUniqueItems(Comparer.Instance))
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return default;
+            }
+
+            UpdateProject(workItem.ProjectKey, workItem.ProjectInfo);
```

Comment:
I would expect this to fail in most cases, since it calls into the `RazorProjectService.UpdateProject`, which requires the dispatcher. However, we're not running on the dispatcher here.

https://github.com/dotnet/razor/blob/07e194a65a1a9f85fec3af041aace521300381b7/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectSystem/RazorProjectService.cs#L241-L251

The easiest solution would be to make `UpdateProject` async, have it call `ProjectSnapshotManagerDispatcher.RunAsync()`, and then await that method here.

Out of curiosity, is this covered by test? I would expect tests to fail with a dispatcher assert here.

Reply:
Now that I've merged my change to make `RazorProjectService` async, you don't even need the dispatcher. You can just call `_projectService.UpdateProjectAsync(...)`

Reply:
Will rebase and update

Reply:
This should be all fixed up, please let me know if I missed anything.